### ISSUE TITLE
housenumbers on building parts aren't rendered yet

### ISSUE THREAD

**Issue by Atalanttore** [gravitystorm/openstreetmap-carto#1746]
Hi,

it would be useful to see the tagged housenumbers (<code>addr:housenumber=XX</code>) of building parts (<code>building:part=yes</code>) on OpenStreetMap.org.

Example of building parts with unrendered housenumbers: http://www.openstreetmap.org/#map=19/49.00980/12.12049

Regards,
Ettore

**Comment by polarbearing** [comment]
```building:part``` is an element from 3D-mapping, that is used to characterise physical differences in the parts, such as different height or materials. It always needs a building outline, carrying the ```building=*``` tag. This building outline is supposed to carry further information, such as address and usage details.

In the example above, there is an outline, and several ```building:part(s)``` that only carry the ```addr:*``` tags. This can be achieved with simple address nodes within the apartment block. 

If these are physically individual buildings wall-to-wall, such as terraced houses, the block can be split, i.e. what are now ```building:part``` can become ```building=house```.

**Comment by dieterdreist** [comment]
sent from a phone

> Am 15.08.2015 um 09:46 schrieb polarbearing <notifications@github.com>:
> 
> building:part is an element from 3D-mapping,
> 


not only 3D


> that is used to characterise physical differences in the parts, such as different height or materials.
> 


not only physical differences, can be all kind of properties like wikipedia links, architectural typology, name, address, usage, ...


> In the example above, there is an outline, and several building:part(s) that only carry the addr:* tags. This can be achieved with simple address nodes within the apartment block.
> 


there is nothing wrong with assigning an address to an area, indeed it has certain advantages over a node

**Comment by pnorman** [comment]
Because the issue is actually about non-building areas, not building parts, I've revised the topic to reflect this.

The relevant SQL is

```sql
The relevant SQL is
```sql
(SELECT
    way,
    "addr:housenumber",
    way_area/NULLIF(!pixel_width!::real*!pixel_height!::real,0) AS way_pixels
  FROM planet_osm_polygon
  WHERE "addr:housenumber" IS NOT NULL
    AND building IS NOT NULL
UNION ALL
SELECT
    way,
    "addr:housenumber",
    NULL AS way_pixels
  FROM planet_osm_point
  WHERE "addr:housenumber" IS NOT NULL
  ORDER BY way_pixels DESC NULLS LAST
) AS housenumbers
```

This code is **old**, coming from [osm.xml SVN revision 15913](https://trac.openstreetmap.org/changeset/15913/subversion), written by ldp in 2009.

The logic is even older, coming from [svn revision 13495](https://trac.openstreetmap.org/changeset/13495/subversion) (see way_numbers), written by Steve Chilton in Feb 2009, 7 years ago.

To see how dropping the building is not null condition would look, I'd want to try it out.

**Comment by polarbearing** [comment]
> there is nothing wrong with assigning an address to an area

There are also amenities, such as amenity=[hospital | school | community_centre], that have one address for the whole campus. In Berlin, the "housenumber" is in fact a number for the plot of land, thus it would be logical that attaching addr:* to the amenity outline gets rendered. 

Consequently, however, we would need to show both number and name, thus #735 would need to be extended to non-buildings.

**Comment by SelfishSeahorse** [comment]
Hi

Are there any plans for this issue?

[According to the wiki](https://wiki.openstreetmap.org/wiki/Addresses), it's fine to add addresses to any areas. Besides, it's more exact than adding them to a node and it doesn't require to duplicate the address for POIs inside that area (e.g. for buildings that are separated inside and thus have more than one house number).

It would be nice to see addresses on non-building areas rendered on OSM Carto.

**Comment by polarbearing** [comment]
We would need to discuss if we compile a positive list of areas on which addr:housenumber is rendered, or do so on any area.

**Comment by dieterdreist** [comment]
2018-01-22 14:35 GMT+01:00 polarbearing <notifications@github.com>:

> We would need to discuss if we compile a positive list of areas on which
> addr:housenumber is rendered, or do so on any area.
>

I would suggest to render them on any area (i.e. consider the housenumber a
feature in this rendering, not a property of something else).

**Comment by kocio-pl** [comment]
Is anybody ready to make the code for testing?

**Comment by polarbearing** [comment]
What about the conflict with names. For buildings, currently, the housenumber is not rendered when the building or the feature using the building has a name. For areas, we have some where the name scales with way-size. 
Both are currently placed in the middle. Would it make sense to place the housenumber more peripheral (e.g. similar to naming borders)? Would it be possible to turn the housenumber towards the street matching the addr:street name?

**Comment by tguen** [comment]
It's common to tag apartment complexes and mobile home parks with a single address, usually (not always) with `landuse=residential`. I would like to see addresses rendered in these cases.

**Comment by hungerburg** [comment]
Here is one building https://www.openstreetmap.org/way/314228700 that has two house numbers, that were put on building:parts. I would appreciate, if the numbers where shown on the map; Even on zoom 18 there is enough space for them not to overlap the name of the building.

**Comment by imagico** [comment]
I would be reluctant to add rendering of housenumbers to polygons with building:part because that does not seem to be a well established tagging at the moment.  The most common way to map different housenumbers on a single building right now is to tag the housenumber on the entrance ([918k combinations](https://taginfo.openstreetmap.org/keys/entrance#combinations) - vs. [20k combinations](https://taginfo.openstreetmap.org/keys/building:part#combinations)) - which tends to be the verifiable information available to the mapper (the housenumber tends to be indicated at the entrance).  Which parts of the building internally correspond to which housenumber is not usually publicly visible.  If or not there are other polygon features other than buildings where tagging housenumbers is common practice is a different matter.

**Comment by hungerburg** [comment]
Sticking the number to the part is nice, so the complex can get a name and the house-numbers still (could) show. Agree, with the numbers on the entrances, this is working right now. And that is where they are available for inspection from the outside. But it is oh so fashionable to map everything as an area. Last year, I mapped the entrances for the parts because they are not immediately obvious. One of the parts has two of them with a full table of door bell buttons. A week ago, I mapped a newly built house close by as two buildings, because there are two addresses. Actually, I could not tell from the outside, where the building parts are separated. It will not get on any aerial this year or next, but that very likely wont help either. It also has a name for the complex, but I did not bother to map that. I thought of it as a marketing gag. So going to tag the entrances. House numbers considered indispensable here :)

**Comment by dieterdreist** [comment]
sent from a phone

> On 5 Nov 2021, at 23:36, hungerburg ***@***.***> wrote:
> 
> A week ago, I mapped a newly built house close by as two buildings, because there are two addresses.


ideally the housenumbers should be added to where they apply, although in Germany for instance this is hardly possible because of the „we do not map plots“ doctrine.
Actually housenumbers a two fold, they are (typically) assigned to something physical like a building, a plot, an entrance, a gate, etc. but then they also serve as addresses for people and things, and we are using the same tags for it. In case the housenumber is for a point, we cannot unambiguously rely on spatial inheritance and repeating the address as property seems the most sane way (relations would be an alternative).

**Comment by aceman444** [comment]
Yes, rendering "name" of "building:part" and also its "addr:*" would be great.

**Comment by endim8** [comment]
This should really be added. I'd imagine the main reason it isn't as widely adopted is _because_ it's not rendered in Carto.

**Comment by aceman444** [comment]
True, because some people may map building parts as separate "building=*" because the name renders. When mapping as a single encompassing building and having building parts with names the names disappear.

**Comment by mistenman** [comment]
Pitkännotkontie 13 E [(view on osm.org)](https://www.openstreetmap.org/way/1559123838) is mapped as a `building:part=house` due to being a subdivision of the larger building structure. It has a clearly defined footprint within the larger structure, and its geometry is accurately mapped. Despite being only a part of the larger building, it is an individual addressable home, which makes its `building:part=house` way the natural feature to tag with address details.

It and its neighbors don't display a housenumber overlaid on top of their polygons in OSM Carto. I suggest that `building:part=*` features be treated equally to `building=*` features when it comes to rendering addresses.

<img height="300" alt="Image" src="https://github.com/user-attachments/assets/bacc4daf-7e68-4d5a-92d6-4a906dfb2763" />
<img height="300" alt="Image" src="https://github.com/user-attachments/assets/2ef6630b-4c75-4625-b9b5-7a1a64485922" />


<sub>ps. don't worry about the hedge on the top left, it's fixed</sub>

**Comment by imagico** [comment]
Principal arguments have already been provided above why this might not be advisable.

Apart from that - situations like the specific case shown as example is elsewhere more commonly mapped with separate buildings (like [here](https://www.openstreetmap.org/#map=17/54.077555/9.958007) or [here](https://www.openstreetmap.org/#map=19/52.418712/20.744080)).

Current numbers:

* `building:part!=no` + `addr:housenumber`: 45k uses
* `entrance=*` +  `addr:housenumber`: 1.8M uses

**Comment by mistenman** [comment]
> Current numbers:
> 
>     * `building:part!=no` + `addr:housenumber`: 45k uses
> 
>     * `entrance=*` +  `addr:housenumber`: 1.8M uses

It's impossible to justify a change of housenumber overlay rendering just for `building:part=*` features due to the minuscule usage, but the title of this issue and the actual overall problem do extend far beyond `building:part=*`. 

For example, in a single industrial complex, the whole area of the complex might have clear boundaries due to a surrounding fence. The address of this complex refers to the whole area in both a legal and what-I-can-see-on-the-ground sense. Tagging the individual buildings within the complex with that same address would, in my interpretation of OSM tagging, be a mistake, as I interpret an address tag on a buildings or `landuse=*` to determine where the address IS. For features such as business POIs, the address tags seem to function as a representation of what is AT an address. 

This is something that's puzzled me since yesterday, and I'm going to go to the forum to seek clarity on the dichotomy before I weigh in with any more suggestions. Edit: discussion at https://community.openstreetmap.org/t/does-addr-represent-where-the-address-is-or-what-is-at-the-address/147772/

**Comment by imagico** [comment]
Our goal is to support mapper in consistent mapping practice, not to nudge mappers to adopt a mapping practice we deem desirable.  Hence the suggestion to just flatly render address labels on polygons which have address tags as primary tags independent of other tagging is not going to work for us, unless mapper adopt this practice at scale in a consistent fashion (which seems unlikely at the moment).

Currently >99 percent of all `addr:housenumber` tags on polygons are in combination with `building=*` - see:

https://taginfo.openstreetmap.org/keys/addr%3Ahousenumber?filter=ways#combinations

the rest are mostly combined with `amenity=*` and `shop=*` on polygons that are not buildings.  For those we would need to introduce label displacement to show both point symbol/label and address - which is a hard problem.

`addr:housenumber` + `landuse=*` comes in with about 230k uses (compared to >90M uses on buildings).

The practice that addresses are tagged both as a primary tag (address nodes with just an address) and secondary tags (on POIs/buildings indicating this feature has the address in question, but is not necessarily the only feature with that address) is not going to go away.

Slightly off-topic (since - as said - we are not in the habit of telling mappers how to map things): If mappers would indeed widely adopt the practice of tagging addresses primarily on non-building polygons as primary tags (in the meaning of *this is the exact area on the ground to which this address applies*) that would be of rather limited use for us to render addresses in a meaningful way.  Think, for example, of cases like

https://www.openstreetmap.org/way/825214435
https://www.openstreetmap.org/way/8922985

To be useful for the map user we would want to display the address at the main entrance or on the administrative building.  You could try to do that by analyzing spatial relationships, but that is highly non-trivial.

Bottom line: Meaningful rendering of addresses tagged as primary tags on polygons is tricky and we would only try to do that if and when mappers actually adopt that mapping paradigm at scale - which so far does not seem to be happening.

**Comment by mistenman** [comment]
Of course, I understand that rendering addresses on arbitrary polygons would create an awfully unreadable map (imagine individually addressed buildings within an area also tagged with an address) unless years of complex rendering logic was to be amended to Carto. It's an unfortunate schism between what works best for renderers and what is most accurate for a database.

**Comment by imagico** [comment]
Not really.  As it has been pointed out in the previous discussion there is an unfortunate misconception widespread among mappers that polygons are an inherently better and more accurate representation of geographic features than linear ways or nodes.

The problem of mappers deciding to use polygons with addresses as primary tags is not that it is complex for us to interpret the data (because it is not), the problem would be that some mappers will dogmatically believe that this is inherently perfect mapping while in reality it is not and you would end up with plenty of meaningless cargo culted polygon geometries providing much less accurate information than address nodes and addresses as secondary tags on buildings and POIs.  Hence we would need to resort to heuristics to attempt to infer cruical information that is not mapped from context.

Anyway - this discussion is theoretical since mappers have not adopted this mapping practice on a larger scale and there are no indications that they are going to.  This issue remains open because we are open to the possibility of rendering address labels on specific non-building polygons as secondary tags (though the need of displacement and de-duplication is a factor in that).

**Comment by aceman444** [comment]
You are taking statistics for the whole world, but there are regional differences supported by laws, not mappers. In some countries it is more common to put addresses onto building:parts or on entrances (multiple addresses in the same building), because the country considers a building a unit too with a single number (not address). Different number for the building (conscription number) and another one for the entrance (street number). Thus your suggestion to just split a building into separate "building" objects (not parts) may be problematic as then you would have duplicate conscription numbers on these objects. As the conscription number is also part of the address, it still needs to be duplicated on parts or entrances, but that is less 'bad' as duplicating it on buildings.

**Comment by imagico** [comment]
Please stay on topic, this is not about conscription numbers (#2016), or - for example - `ref` on entrances (#4687).

And i did not suggest mappers to map things a certain way, in fact i explicitly said that we do not want to nudge mappers to a certain mapping practice.  Please read carefully before making such claims.  What i pointed out is that the specific case example given by @mistenman differs from the most common mapping practice for such cases.

Finland, where the case example given is from, has 534 combinations of `building:part` and `addr:housenumber` on ways/relations - which is proportionally somewhat more than world wide, but not much.  A lot of these are redundant, i.e. several building parts of the same building carrying the same address or the building part carrying the same address as the building it is in.

Generally speaking we produce a map style for a global audience, we try to do our best to represent the global *diversity in geography* (in the sense of also displaying things that only exist in some part of the world) but we explicitly do not want to support regional *diversity in mapping practice* without a basis in the observable geography (local mappers deciding to map the same things differently in their area than what is global consensus).  If you feel that the current mapping practice is bad then you need to try to convince the mappers to change it.  If and when you manage to do that and this becomes visible in the data on a global scale then we can discuss adjusting rendering here.

**Comment by aceman444** [comment]
Well, you replied to him that his case can be mapped with individual buildings, in comment https://github.com/openstreetmap-carto/openstreetmap-carto/issues/1746#issuecomment-5698680735 . It may not mean you explicitly "suggest" it, but is not far from it.
So while others do it differently, it does not mean his way is invalid or should be reworked. It may be both ways are correct.
I can also play with your verbatim words. I did not say anything about 'ref' on entrances so why are you referencing them. I meant putting addr:* tags on entrances.
I didn't even mean anything about rendering conscription numbers, just meaning the concept of it and that people may want to keep mapping of 1 conscription number=1 OSM building and using building:parts for any subdivisions, whether differing 3D properties, names or addresses.
So please try to understand the general idea people are meaning, not their exact spelling (there may be different language backgrounds involved too).

But back on topic. There are also other objects having an address. Aside from the mentioned cases like hospital campuses, industrial areas, or amenity=school areas, which may have a main address on that campus polygon, but individual buildings inside may have different addresses. Some countries also assign addresses to other objects that are not necessarily buildings, but are man made constructions, e.g. industrial works (like pipes and processing chambers in an oil refinery), gasometers, pools, large wastewater treatment reservoirs, etc. So it could be considered if these should be rendered.

**Comment by mistenman** [comment]
Rendering housenumbers on larger features like hospital campuses is probably tricky since it's ambiguous where the number should be placed for maximum clarity. I think putting it next to the street like this map does would be a good solution to consider, if it is ever to be implemented. Only works if the area with the address is next to the street, of course.

<img width="679" height="686" alt="Image" src="https://github.com/user-attachments/assets/537f250a-5eba-47fe-9d3b-084730b126c6" />
